Additional sbt commands to build stdlib packages

hubertp opened the pull request:
[ci no changelog needed]

### Pull Request Description

Auxiliary sbt commands for building individual
stdlib packages.
The commands check if the engine distribution was built at least once,
and only copy the necessary package files if necessary.
So far added:
- `buildStdLibBase`
- `buildStdLibDatabase`
- `buildStdLibTable`
- `buildStdLibImage`
- `buildStdLibGoogle_Api`

Related to [#182014385](https://www.pivotaltracker.com/story/show/182014385)

### Checklist

Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

hubertp committed:
Additional sbt commands to build stdlib packages

Initial draft to illustrate aux sbt commands for building individual
stdlib packages.
The commands check if the engine distribution was built at least once,
and only copy the necessary package files if necessary.
So far added:
- `buildStdLibBase`
- `buildStdLibDatabase`
- `buildStdLibTable`
- `buildStdLibBaseVis`

radeusgd commented on build.sbt:
```diff
@@ -1737,6 +1737,76 @@ buildEngineDistribution := {
+  )
+  if (result) {
+    log.info(s"Package '$name' has been updated")
+    DistributionPackage.fixLibraryManifest(targetPkgRoot, version, log)
```
The manifest is only needed for distribution, it does not affect anything during regular runtime.
Generating the manifest unfortunately takes some time as we need to analyze the imports and currently our parser is quite slow.
Since, if I understand correctly, these commands are meant to be just convenience methods for development, because we still are using `buildEngineDistribution` for distribution, I think maybe we could skip this step? It should cut on the build time quite a bit and doesn't seem needed.
I didn't do measurements but from practice I think it often takes more time than the incremental engine JAR build (because if there were no modifications it just checks that and quits quite quickly).

hubertp committed:
Add associated java code to packages

Some packages in stdlib have associated java code. That also has to be
updated.

hubertp committed:
Generate manifests

hubertp committed:
Merge branch 'develop' into wip/hubert/build-individual-library-components-182014385

jdunkerley reviewed (approved):
Looks good.

There are a couple of other libraries but we can add those when we are working on them more full time.

hubertp committed:
Merge branch 'develop' into wip/hubert/build-individual-library-components-182014385

mwu-tow reviewed (commented):
1. Why are we introducing a separate command for each library, rather than defining a single parametrized command like `buildStdLib <name>`?
2. The implementation code is heavily duplicated. Can it be somehow refactored?
3. In any case, the new commands should be documented somewhere.

hubertp commented:
> 1. Why are we introducing a separate command for each library, rather than defining a single parametrized command like `buildStdLib <name>`?
> 2. The implementation code is heavily duplicated. Can it be somehow refactored?

I can do that but then command completion won't work

radeusgd commented on project/DistributionPackage.scala:
```diff
@@ -42,7 +42,7 @@ object DistributionPackage {
     source: File,
     destination: File,
     cache: CacheStore
-  ): Unit = {
+  ): Boolean = {
```
I'd suggest documenting what each return value means.

radeusgd reviewed (approved):
Looks good, but as @mwu-tow mentioned - there's still a significant amount of duplication. Cannot we reduce it a bit?

hubertp committed:
Use InputTask with custom Parser to avoid duplication

1) conditional inside the task requires delegation to another task
2) couldn't find an easy way to generalize compilation task for
individual std-bits so that part remains hardcoded (might be an exercise
for future)

hubertp committed:
Merge branch 'develop' into wip/hubert/build-individual-library-components-182014385

hubertp committed:
more docs

hubertp commented:
@radeusgd @mwu-tow OK, so that's as far as I'm willing to push it today without throwing a monitor out of the window (did I mention how much I love sbt?). That significantly reduces the duplication.

hubertp commented:
And yes, you get auto-completion with the custom parser I added. I was wrong about that one :)

hubertp committed:
Merge branch 'develop' into wip/hubert/build-individual-library-components-182014385

hubertp commented on project/StdBits.scala:
```diff
@@ -113,4 +115,31 @@ object StdBits {
+    if (!sourceDir.exists) {
+      throw new RuntimeException("Invalid standard library package " + name)
+    }
+    LibraryManifestGenerator.generateManifests(
```
Note that this and `fixLibraryManifest` are essentially copy pasted from `buildEngineDistribution` task because I assumed this was needed.

radeusgd commented on project/StdBits.scala:
```diff
@@ -113,4 +115,31 @@ object StdBits {
+      file("distribution"),
+      log,
+      cacheFactory
+    )
```
I'm actually not sure if this is needed.

The manifest generation takes some time (because our parser is _sloooow_). OTOH the data from the manifest is only important for downloading libraries from a repository. It's not important when the libraries are used as we do use them for testing.

So as long as we keep using these particular commands just for development and keep using the full `buildEngineDistribution` which performs the manifest generation for distribution, I think we should be fine to remove this part. And that would make the process a bit faster, which I think may be desirable.

hubertp replied:
sgtm, I will remove them then

hubertp replied:
done

hubertp committed:
Remove obsolete tasks as requested

+ more docs

hubertp committed:
Merge branch 'develop' into wip/hubert/build-individual-library-components-182014385

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/build-individual-library-components-182014385
